### PR TITLE
Add finger painting script

### PR DESCRIPTION
**Pull request by ctrlaltdavid**

### CODE REVIEW

**Comment by ctrlaltdavid** [comment]
Testing in HMD view:
- A tablet "PAINT" button toggles finger painting on/off.
- Controller trigger paints red polylines from index finger tip as you move your hand.
- Index finger is pointed straight while painting.
- Line width varies with degree of trigger press.
- Controller grip press erases nearest polyline within 0.1m of finger tip.
- Can paint and erase with both hands simultaneously.
- Lasers and grabbing are disabled when "PAINT" is on and the tablet is closed.

Known issue: The tablet button uses the same graphic as the "Bubble" button; this graphic will be updated.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3712aea8613435fa1a04483d1f3c77836f93a249 available at: https://gist.github.com/910044952c37d9b4013da3027af01766<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9696-3712aea.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9696-3712aea.dmg

**Comment by PhilipRosedale** [comment]
Works great!  Amazing great that you turned off the beam with a message and pointed the finger.  Two things I noticed: 

1.  Sometimes when painting while pointing the hand upwards,  the beam briefly comes back on.  Bug?

2.  When drawing slowly, the creation of new segments seems to chunky... can we make the creation of a new segment a function of velocity or something?   I'd like to see segments appear at the same rate regardless of whether I am moving slowly or fast.

**Comment by PhilipRosedale** [comment]
Looking at the code, I think a solution would be to make MIN_STROKE_LENGTH something quite small, like a few millimeters, but also don't draw a new segment more than 15 times a second or so.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0689557ba57d64516522dc8f4a1c2464c005d6aa available at: https://gist.github.com/910044952c37d9b4013da3027af01766<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9696-0689557.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9696-0689557.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2c02b42c967354be7dd559c67e8a0a05768a7d5b available at: https://gist.github.com/910044952c37d9b4013da3027af01766<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9696-2c02b42.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9696-2c02b42.dmg

**Comment by ctrlaltdavid** [comment]
@PhilipRosedale 
1) The laser beams appearing sometimes was when pointing at, I believe, the (invisible) old toolbar. I've n now disabled the pointer while painting.
2) I've improved slow drawing per your suggestions; I think it's pretty good now? Note: Polylines have a hard-coded limit of a maximum of 70 points.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 692fa52acaa3a8d1debeefa6932bd86907675613 available at: https://gist.github.com/910044952c37d9b4013da3027af01766<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9696-692fa52.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9696-692fa52.dmg

**Comment by PhilipRosedale** [comment]
Works great.  Drew with my daughter for an hour!

File: scripts/system/fingerPaint.js [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+            setUp: setUp,
+            tearDown: tearDown
+        };
+    }
+
+    function updateHandFunctions() {
+        // Update other scripts' hand functions.
+        var enabled = !isFingerPainting || isTabletDisplayed;
+
+        Messages.sendMessage(HIFI_GRAB_DISABLE_MESSAGE_CHANNEL, JSON.stringify({
```
Comment (ZappoMan) [inline scripts/system/fingerPaint.js]: why are these messages being sent to all clients in the domain? is that desired? Did you mean for these to be local only?

Comment (ctrlaltdavid) [inline scripts/system/fingerPaint.js]: Thanks!  Fixed.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 71ba38afb05c3712b3fc34a52b7851b41ec21fbc available at: https://gist.github.com/910044952c37d9b4013da3027af01766<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9696-71ba38a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9696-71ba38a.dmg

**Comment by ctrlaltdavid** [comment]
Test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 71ba38afb05c3712b3fc34a52b7851b41ec21fbc available at: https://gist.github.com/910044952c37d9b4013da3027af01766<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9696-71ba38a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9696-71ba38a.dmg

**Comment by ZappoMan** [comment]
@ctrlaltdavid @PhilipRosedale - I removed QA Approved because it looks like David made some significant changes to the enable logic, and so I think this deserves to be tested again.

File: scripts/system/fingerPaint.js (hunk 1 above)
Comment (ZappoMan) [inline scripts/system/fingerPaint.js]: 👍

**Comment by Nex-Pro** [comment]
@ZappoMan  this is still showing as QA ready.

**Comment by ZappoMan** [comment]
@nex-pro it needs a retest

**Comment by Nex-Pro** [comment]
@ZappoMan  My bad Just woke up after an all-night coding session. Mis-read. BTW good job Friday night.

Will test this.

**Comment by Nex-Pro** [comment]
@ZappoMan @ctrlaltdavid  I'll test this with Vive  Should we get tests with Touch ?

**Comment by Nex-Pro** [comment]
@PhilipRosedale @ctrlaltdavid  I'd like to design the Icon for finger paint. I put a job up on worklist.
https://worklist.net/21184

**Comment by Nex-Pro** [comment]
### SWEET

Works well. methinks this is a pass.

Painting: Really good fluid and reasonably predictable. See images.

Erasing.
I made a 3d array of small strokes in a cube about 30 cm on a side maybe 25mm apart 
and was able to erase each one even though they were ( using squeeze wasn't the way I'd go ) 
closely surrounded by neighbours.


#### FEEDBACK

I hope there will be some further development of this.
Would be nice if the ends of the strokes didn't cut off abruptly in a straight line.
Also how about it continues to add strokes as long as the trigger is held say up to a pre-set max number then warn user with some haptic pulses that the ink is about to run out.
Not keen on the grip to erase. ( maybe OK if continuous after a set time ).
Needs more eraser options.


hifi-snap-by--on-2017-02-19_18-06-15
hifi-snap-by--on-2017-02-19_18-09-58

**Comment by Nex-Pro** [comment]
@ZappoMan @ctrlaltdavid  Just noticed that the polylines default to -1 lifetime. ) I went back to dev-playa.

**Pull request merged**